crypto/evp/kdf_meth.c

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/evp.h>
+#include <openssl/err.h>
+#include <openssl/core.h>
+#include <openssl/core_numbers.h>
+#include <openssl/kdf.h>
+#include "internal/evp_int.h"
+#include "internal/provider.h"
+#include "evp_locl.h"
+
+static int evp_kdf_up_ref(void *vkdf)
+{
+    EVP_KDF *kdf = (EVP_KDF *)vkdf;
+    int ref = 0;
+
+    CRYPTO_UP_REF(&kdf->refcnt, &ref, kdf->lock);
+    return 1;
+}
+
+static void evp_kdf_free(void *vkdf){
+    EVP_KDF *kdf = (EVP_KDF *)vkdf;
+    int ref = 0;
+
+    if (kdf != NULL) {
+        CRYPTO_DOWN_REF(&kdf->refcnt, &ref, kdf->lock);
+        if (ref <= 0) {
+            ossl_provider_free(kdf->prov);
+            OPENSSL_free(kdf->name);
+            CRYPTO_THREAD_lock_free(kdf->lock);
+            OPENSSL_free(kdf);
+        }
+    }
+}
+
+static void *evp_kdf_new(void)
+{
+    EVP_KDF *kdf = NULL;
+
+    if ((kdf = OPENSSL_zalloc(sizeof(*kdf))) == NULL
+        || (kdf->lock = CRYPTO_THREAD_lock_new()) == NULL) {
+        evp_kdf_free(kdf);
```

Comment:
is this call ok if kdf->lock is NULL?

Reply:
It looks like it is.  Any particular concerns?

Reply:
DOWNREF on a NULL lock was what I am referring to..

Reply:
Okay, I've just made it an OPENSSL_free instead.